File: apps/party-time-backend/src/main/java/com/partytime/service/AddressService.java
package com.partytime.service;

import com.partytime.api.dto.address.AddressDTO;
import com.partytime.api.error.ApiError;
import com.partytime.jpa.entity.Address;
import com.partytime.jpa.mapper.AddressMapper;
import com.partytime.jpa.repository.AddressRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class AddressService {

    private final AddressRepository addressRepository;

    public Address getAddress(String addressLine,
                              String addressLineAddition,
                              String zip,
                              String city,
                              String country) {
        String adLineAdd = addressLineAddition != null ? addressLineAddition : "";
        return addressRepository.findByData(addressLine, adLineAdd, zip, city, country)
            .or(() -> Optional.of(AddressMapper.create(addressLine, adLineAdd, zip, city, country))
                .map(addressRepository::save))
            .orElseThrow(() -> ApiError.badRequest("Fehler bei der Anlage einer Addresse").asException());
    }

    public Address getAddress(AddressDTO address) {
        return getAddress(address.getAddressLine(),
            address.getAddressLineAddition(),
            address.getZip(),
            address.getCity(),
            address.getCountry());
    }

}

File: apps/party-time-backend-kotlin/src/main/kotlin/com/partytime/service/AddressService.kt
package com.partytime.service

import com.partytime.api.dto.address.AddressDTO
import com.partytime.jpa.entity.Address
import com.partytime.jpa.repository.AddressRepository
import org.springframework.stereotype.Service

/**
 * A [Service] class for address related functionality.
 *
 * @param addressRepository A repository where addresses can be stored
 * @constructor Constructs a new [AddressService]
 */
@Service
class AddressService(
    private val addressRepository: AddressRepository
) {

    /**
     * Implemented during F001
     *
     * Saves a new address to the database if it has not been saved before.
     *
     * @param addressLine Address line containing the street name and house number
     * @param addressLineAddition Additional (optional) address line
     * @param zip Postal code of the city
     * @param city Name of the city
     * @param country Name of the country
     * @return A newly stored [Address] or a matching [Address] that has been stored before
     */
    fun saveAddress(
        addressLine: String,
        addressLineAddition: String?,
        zip: String,
        city: String,
        country: String
    ): Address = addressRepository.findByData(addressLine, addressLineAddition, zip, city, country)
        .orElseGet {
            Address(
                addressLine = addressLine,
                addressLineAddition = addressLineAddition,
                zip = zip,
                city = city,
                country = country
            ).let(addressRepository::save)
        }

    /**
     * Saves the address stored in an [AddressDTO].
     *
     * @param address Address information to be saved
     * @return A newly stored [Address] or a matching [Address] that has been stored before
     */
    fun saveAddress(address: AddressDTO) = saveAddress(
        address.addressLine,
        address.addressLineAddition,
        address.zip,
        address.city,
        address.country
    )
}
